File: frontend/src/api/notifications.ts
import { getList, post } from "@/api/client";
import type { Paginated } from "@/api/client";
import type { Notification } from "@/types/api";

export const notificationsApi = {
  async list(unreadOnly = false): Promise<Paginated<Notification>> {
    return getList<Notification>("/notifications/", { params: unreadOnly ? { unread_only: "true" } : undefined });
  },
  async markRead(notificationId: string): Promise<Notification | null> {
    return post<Notification | null>(`/notifications/${notificationId}/read/`);
  },
  async markAllRead(): Promise<{ updated: number }> {
    return post<{ updated: number }>("/notifications/read-all/");
  },
};

File: backend/app/routers/notifications.py
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy import select
from sqlalchemy.orm import Session

from ..database import get_db
from ..dependencies import current_user
from ..models import Notification, User
from ..schemas import NotificationOut

router = APIRouter(prefix="/notifications", tags=["notifications"])


@router.get("", response_model=list[NotificationOut])
def list_notifications(unread_only: bool = False, user: User = Depends(current_user), db: Session = Depends(get_db)):
    query = select(Notification).where(Notification.user_id == user.id)
    if unread_only:
        query = query.where(Notification.is_read.is_(False))
    notifications = db.scalars(query.order_by(Notification.created_at.desc())).all()
    return notifications


@router.get("/unread-count")
def unread_count(user: User = Depends(current_user), db: Session = Depends(get_db)):
    total = len(db.scalars(select(Notification).where(Notification.user_id == user.id, Notification.is_read.is_(False))).all())
    return {"unread": total}


def _owned_notification(notification_id: int, user: User, db: Session) -> Notification:
    notification = db.get(Notification, notification_id)
    if not notification or notification.user_id != user.id:
        raise HTTPException(404, "Notification not found")
    return notification


@router.post("/{notification_id}/read", response_model=NotificationOut)
def mark_read(notification_id: int, user: User = Depends(current_user), db: Session = Depends(get_db)):
    notification = _owned_notification(notification_id, user, db)
    notification.is_read = True
    db.commit()
    db.refresh(notification)
    return notification


@router.post("/read-all")
def mark_all_read(user: User = Depends(current_user), db: Session = Depends(get_db)):
    notifications = db.scalars(select(Notification).where(Notification.user_id == user.id, Notification.is_read.is_(False))).all()
    for notification in notifications:
        notification.is_read = True
    db.commit()
    return {"updated": len(notifications)}
